yarn-project/prover-client/src/prover/bb_prover.ts

```diff
@@ -233,27 +282,110 @@ export class BBNativeRollupProver implements CircuitProver {
       throw new Error(provingResult.reason);
     }
 
+    // Ensure our vk cache is up to date
+    await this.updateVerificationKeyAfterProof(provingResult.vkPath!, circuitType);
+
     // Read the proof and then cleanup up our temporary directory
-    const proofBuffer = await fs.readFile(provingResult.path!);
+    const proof = await fs.readFile(`${provingResult.proofPath!}/${PROOF_FILENAME}`);
+
+    await fs.rm(bbWorkingDirectory, { recursive: true, force: true });
+
+    logger.info(`Generated proof for ${circuitType} in ${provingResult.duration} ms, size: ${proof.length} fields`);
+
+    return [outputWitness, new Proof(proof)];
+  }
+
+  /**
+   * Executes a circuit and returns it's outputs and corresponding proof with embedded aggregation object
+   * @param witnessMap - The input witness
+   * @param circuitType - The type of circuit to be executed
+   * @param convertOutput - Function for parsing the output witness to it's corresponding object
+   * @returns The circuits output object and it's proof
+   */
+  public async createRecursiveProof<PROOF_LENGTH extends number, CircuitOutputType>(
+    witnessMap: WitnessMap,
+    circuitType: ServerProtocolArtifact,
+    convertOutput: (outputWitness: WitnessMap) => CircuitOutputType,
+  ): Promise<[CircuitOutputType, RecursiveProof<PROOF_LENGTH>]> {
+    // Create random directory to be used for temp files
+    const bbWorkingDirectory = `${this.config.bbWorkingDirectory}/${randomBytes(8).toString('hex')}`;
+    await fs.mkdir(bbWorkingDirectory, { recursive: true });
+
+    await fs.access(bbWorkingDirectory);
+
+    // Have the ACVM write the partial witness here
+    const outputWitnessFile = `${bbWorkingDirectory}/partial-witness.gz`;
+
+    // Generate the partial witness using the ACVM
+    // A further temp directory will be created beneath ours and then cleaned up after the partial witness has been copied to our specified location
+    const simulator = new NativeACVMSimulator(
+      this.config.acvmWorkingDirectory,
+      this.config.acvmBinaryPath,
+      outputWitnessFile,
+    );
+
+    const artifact = ServerCircuitArtifacts[circuitType];
+
+    logger.debug(`Generating witness data for ${circuitType}`);
+
+    const outputWitness = await simulator.simulateCircuit(witnessMap, artifact);
+
+    const outputType = convertOutput(outputWitness);
+
+    // Now prove the circuit from the generated witness
+    logger.debug(`Proving ${circuitType}...`);
+
+    const provingResult = await generateProof(
+      this.config.bbBinaryPath,
+      bbWorkingDirectory,
+      circuitType,
+      artifact,
+      outputWitnessFile,
+      logger.debug,
+    );
+
+    if (provingResult.status === BB_RESULT.FAILURE) {
+      logger.error(`Failed to generate proof for ${circuitType}: ${provingResult.reason}`);
+      throw new Error(provingResult.reason);
+    }
+
+    // Ensure our vk cache is up to date
+    await this.updateVerificationKeyAfterProof(provingResult.vkPath!, circuitType);
```

Comment:
hm, do we have to `await` this call? I think it would be safe to let this run async because the cache is updated sync and every call to `verifyProof` awaits the promise read from the cache